File: swift/SupermarketReceipt/SupermarketReceipt/Teller.swift
public class Teller {

    private let catalog: SupermarketCatalog
    private var offers = [Product: Offer]()

    public init(catalog: SupermarketCatalog) {
        self.catalog = catalog
    }

    public func addSpecialOffer(offerType: SpecialOfferType, product: Product, argument: Double) {
        self.offers[product] = Offer(offerType: offerType, product: product, argument: argument)
    }

    public func checksOutArticlesFrom(theCart: ShoppingCart) -> Receipt {
        var receipt = Receipt()
        var productQuantities = theCart.items
        for pq in productQuantities {
            var p = pq.product
            var quantity = pq.quantity
            var unitPrice = self.catalog.getUnitPrice(product: p)
            var price = quantity * unitPrice
            var priceTo3dp = round(100 * price) / 100
            receipt.addProduct(p: p, quantity: quantity, price: unitPrice, totalPrice: priceTo3dp)
        }
        theCart.handleOffers(receipt: receipt, offers: self.offers, catalog: self.catalog)

        return receipt
    }

}

File: python_pytest/src/teller.py
from model_objects import Offer
from receipt import Receipt


class Teller:

    def __init__(self, catalog):
        self.catalog = catalog
        self.offers = {}

    def add_special_offer(self, offer_type, product, argument):
        self.offers[product] = Offer(offer_type, product, argument)

    def checks_out_articles_from(self, the_cart):
        receipt = Receipt()
        product_quantities = the_cart.items
        for pq in product_quantities:
            p = pq.product
            quantity = pq.quantity
            unit_price = self.catalog.unit_price(p)
            price = quantity * unit_price
            receipt.add_product(p, quantity, unit_price, price)

        the_cart.handle_offers(receipt, self.offers, self.catalog)

        return receipt

    def product_with_name(self, name):
        return self.catalog.products.get(name, None)